spec/src/main/asciidoc/configsources.asciidoc

```diff
@@ -157,7 +157,10 @@ Alternatively, for the injection style, the following lookup should be used to r
     @ConfigProperty(name="myprj.some.dynamic.timeout", defaultValue="100")
     private javax.inject.Provider<Long> timeout; 
 ----
-
+Whether a `ConfigSource` supports this dynamic behavior or not is dependent on how it's implemented.
+For an instance, the default `ConfigSource` microprofile-config.properties and Environment Variables are not dynamic 
+while System Properties are dynamic by nature. MicroProfile Config Implementation can decide whether 
```

Comment:
"MicroProfile Config Implementation [sic] can decide whether a `ConfigSource` can be dynamic or not": I find this hard to understand.  `ConfigSource`s can be packaged with applications.  How can a given MicroProfile Config implementation "decide" whether a `ConfigSource` packaged with an application is dynamic or not?

Reply:
It somewhat depends on whether a `ConfigSource` implementation using a backing source that could be updated.

Reply:
@kenfinnigan Not trying to be difficult, but again: how is, say, an OpenLiberty MicroProfile Config implementation shipped with an application server supposed to "know" that my `ConfigSource` implementation—the one in my application deployed to that server—is dynamic?  It cannot.

Reply:
Ah, see your point now.

So yes, an MP Config implementation can only "know" whether the `ConfigSource`s it provides are dynamic or not. There is no way for it to "know" about any `ConfigSource` that is included within the application code directly

Reply:
Correct.  In other words: what is the contract of [`ConfigSource#getPropertyNames()`](https://javadoc.io/static/org.eclipse.microprofile.config/microprofile-config-api/1.3/org/eclipse/microprofile/config/spi/ConfigSource.html#getPropertyNames--)?  It is so underspecified as to be almost useless, _viz._ is the returned `Set` an immutable snapshot?  A mutating snapshot?  Is it thread safe (i.e. what must I do to iterate over it)?  If I call the method two times must I get back equal `Set`s?  Identical `Set`s?  And so on.  (And that's before even considering the [`ConfigSource#getProperties()`](https://javadoc.io/static/org.eclipse.microprofile.config/microprofile-config-api/1.3/org/eclipse/microprofile/config/spi/ConfigSource.html#getProperties--) method, which suffers from the same kind of lack of specificity.)

Reply:
I would agree that the way it's documented, and the fluidity around whether a `ConfigSource` support dynamic updates or not, means that it's impossible to write meaningful TCK tests to verify any of this as it's undefined

Reply:
Right, and not just TCK tests: this is a public API exposed to the end user.  She cannot use it because she does not know how to use the returned `Set`.  Must she synchronize on it?  Or not?  Can she cache it?  And so on.  We are in agreement here at a high level that there is _much_ more work to be done here, I think.

Reply:
I agree that there is still lots to be done.

In terms of this PR, I'm wondering if we need to go back to discussing improving the definitions of the API first.

Reply:
@ljnelson I was going to add a PR to say the getPropertyNames() can be iterated in a thread safe manner based on the suggestion from @dmlloyd 
@kenfinnigan there is a TCK to verify the dynamic aspect of system property.

Reply:
Yes, that's a good start.  An even better start would be to make sure that the iteration is _fully_ specified.  See the [Concurrent Collections section in the package description of the `java.util.concurrent` package documentation](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/package-summary.html#package.description) for a particularly good example.

Reply:
Follow up on what I mentioned there are tcks to assert dynamic config lookup.
ConfigProviderTest.testDynamicValueInPropertyConfigSource()
ConfigProviderTest.testGetPropertyNames()
ConfigProviderTest.testPropertyConfigSource() 
EmptyValuesTest.testEmptyStringProgrammaticLookup()